feat(expo): add support for native css (#40342)
expo/expo: docs/pages/guides/css.mdx

---
title: Functional CSS
sidebar_title: CSS
maxHeadingDepth: 4

> description: Learn about using functional CSS to style native apps

Reviewer: ```suggestion
description: Learn about using functional CSS to style native apps.
```

Missing period.

---

> > Experimentally available in **SDK 55 and above**. This is a beta release and subject to breaking changes.

Reviewer: ```suggestion
> Experimentally available in **SDK 55 and later**. This is a beta release and is subject to breaking changes.
```

Writing style guide instructions on how to use the wording here for SDK version range calluots: https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#sdk-version-range-callouts

Expo's bundler supports full comprehensive CSS for web platforms, but this traditionally doesn't work for React Native because the native platform doesn't support CSS files. However, you can use **Functional CSS** to enable a subset of CSS features on native platforms.

## What is Functional CSS?

Functional CSS is a system for parsing CSS styles to JS objects, then using React hooks to apply those objects to components and reactively update them when media queries and CSS variables change. This is limited however:

- Only class selectors are supported (e.g. `.my-class { color: red; }`), element selectors (e.g. `div { color: red; }`) and ID selectors (e.g. `#my-id { color: red; }`) are not supported.
- Only a subset of CSS properties are supported. If the underlying native platform doesn't support a style property, it won't work in Functional CSS.
- Cascading is not supported. This is with exception to nested text elements.

## How to use Functional CSS

> Enable the experimental Functional CSS feature in your **app.json** or **app.config.js**:

Reviewer: ```suggestion
Enable the experimental Functional CSS feature in your [app config](/workflow/configuration/) file:
```

An app config file can also be `app.config.ts`. So let's update it and link its doc here: https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#referencing-appjsonappconfigjsonappconfigjsappconfigts

```json app.json
{
  "experiments": {
    "functionalCSS": true
  }
}
```

Define some CSS styles. You can use CSS Modules, Sass, SCSS, PostCSS, Tailwind CSS, or global CSS styles.

```css globals.css
:root {
  --background: #fff;
  --text: #000;
}

.container {
  flex: 1;
  justify-content: center;
  align-items: center;
  background-color: var(--background);
}

.text {
  font-size: 20px;
  color: var(--text);
}
```

Now wrap components with hooks to convert the `className` prop to a style object:

```tsx app/index.tsx
import { Text as RNText, TextProps, View as RNView, ViewProps } from 'react-native';
import { useCssElement } from 'expo/css';

import '../globals.css';

function Text(props: TextProps & { className?: string }) {
  return useCssElement(RNText, props, {
    className: 'style',
  });
}

function View(props: ViewProps & { className?: string }) {
  return useCssElement(RNView, props, {
    className: 'style',
  });
}

export default function Page() {
  return (
    <View className="container">
      <Text className="text">Index</Text>
    </View>
  );
}
```

> Now run `npx expo` to start the project. You should see a white screen with black text. On web, this will fallback to standard CSS behavior.

Reviewer: ```suggestion
Now run `npx expo start` command in a terminal window to start the project. You should see a white screen with black text. On web, this will fall back to standard CSS behavior.
```

## Platform specification

CSS supports platform extensions such as `globals.ios.css` and `globals.android.css` to define platform-specific styles.

## Media queries

You can use media queries to adapt styles based on the device's screen size, and color scheme. For example, you can add the following CSS to your **globals.css** file:

```css globals.css
@media (prefers-color-scheme: dark) {
  :root {
    --background: #000;
    --text: #fff;
  }
}
```

## Benefits of Functional CSS

- CSS is backwards compatible and follows a well defined specification.
- When you use CSS, Expo can automatically optimize the styles for different platforms resulting in more performant bundles and faster load times.
- CSS has dynamic functionality that would normally require React hooks. React hooks are limiting as they don't support server rendering, and you cannot call them conditionally meaning you need to refactor components to optimize them.
- High-performance animations and built-in. We use the native Reanimated animation library on native and CSS as-is on web. This results in smooth max-fps animations on all platforms.

> - Additional CSS units such as `vw`, `vh`, `%`, `em`, and `rem` are supported.

Reviewer: ```suggestion
- CSS is backwards compatible and follows a well defined specification.
- When you use CSS, Expo can automatically optimize the styles for different platforms, resulting in more performant bundles and faster load times.
- CSS has dynamic functionality that would normally require React hooks. React hooks are limited as they don't support server rendering, and you cannot call them conditionally. This means you need to refactor components to optimize them.
- High-performance animations are built in. We use the native Reanimated animation library on native and CSS as-is on web. This results in smooth max-FPS animations on all platforms.
- Additional CSS units such as `vw`, `vh`, `%`, `em`, and `rem` are supported.
```
